src/gpuarray_buffer_cuda.c

```diff
@@ -937,41 +1021,123 @@ static void *call_compiler(const char *src, size_t len, const char *arch_arg,
 
   err = nvrtcGetPTXSize(prog, &buflen);
   if (err != NVRTC_SUCCESS) goto end;
 
-  buf = malloc(buflen);
-  if (buf == NULL) {
-    nvrtcDestroyProgram(&prog);
-    FAIL(NULL, GA_MEMORY_ERROR);
+  if (strb_ensure(ptx, buflen) == 0) {
+    err = nvrtcGetPTX(prog, ptx->s+ptx->l);
+    if (err == NVRTC_SUCCESS) ptx->l = buflen;
   }
 
-  err = nvrtcGetPTX(prog, (char *)buf);
-  if (err != NVRTC_SUCCESS) goto end;
-
-  *bin_len = buflen;
-
 end:
   nvrtcDestroyProgram(&prog);
-  if (err != NVRTC_SUCCESS) {
-    free(buf);
-    FAIL(NULL, GA_SYS_ERROR);
+  if (err != NVRTC_SUCCESS)
+    return GA_SYS_ERROR;
+  return GA_NO_ERROR;
+}
+
+static int make_bin(cuda_context *ctx, const strb *ptx, strb *bin, strb *log) {
+  char info_log[2048];
+  char error_log[2048];
+  void *out;
+  size_t out_size;
+  CUlinkState st;
+  CUjit_option cujit_opts[] = {
+    CU_JIT_INFO_LOG_BUFFER_SIZE_BYTES,
+    CU_JIT_INFO_LOG_BUFFER,
+    CU_JIT_ERROR_LOG_BUFFER_SIZE_BYTES,
+    CU_JIT_ERROR_LOG_BUFFER,
+    CU_JIT_LOG_VERBOSE,
+    CU_JIT_GENERATE_DEBUG_INFO,
+    CU_JIT_GENERATE_LINE_INFO,
+  };
+  void *cujit_opt_vals[] = {
+    (void *)sizeof(info_log), info_log,
+    (void *)sizeof(error_log), error_log,
+#ifdef DEBUG
+    (void *)1, (void *)1, (void *)1
+#else
+    (void *)0, (void *)0, (void *)0
+#endif
+  };
+  int err = GA_NO_ERROR;
+
+  ctx->err = cuLinkCreate(sizeof(cujit_opts)/sizeof(cujit_opts[0]),
+                          cujit_opts, cujit_opt_vals, &st);
+  if (ctx->err != CUDA_SUCCESS)
+    return GA_IMPL_ERROR;
+  ctx->err = cuLinkAddData(st, CU_JIT_INPUT_PTX, ptx->s, ptx->l,
+                           "kernel code", 0, NULL, NULL);
+  if (ctx->err != CUDA_SUCCESS) {
+    err = GA_IMPL_ERROR;
+    goto out;
   }
-  return buf;
+  ctx->err = cuLinkComplete(st, &out, &out_size);
+  if (ctx->err != CUDA_SUCCESS) {
+    err = GA_IMPL_ERROR;
+    goto out;
+  }
+  strb_appendn(bin, out, out_size);
+out:
+  cuLinkDestroy(st);
+  strb_appends(log, "Link info log::\n");
+  strb_appends(log, info_log);
+  strb_appends(log, "\nLink error log::\n");
+  strb_appends(log, error_log);
+  strb_appendc(log, '\n');
+  return err;
+}
+
+static int compile(cuda_context *ctx, strb *src, strb* bin, strb *log) {
+  strb ptx = STRB_STATIC_INIT;
+  strb *cbin;
+  kernel_key k;
+  kernel_key *pk;
+  int err;
+
+  memcpy(k.bin_id, ctx->bin_id, 64);
+  memcpy(&k.src, src, sizeof(strb));
+
+  // Look up the binary in the disk cache
+  if (ctx->disk_cache) {
+    cbin = cache_get(ctx->disk_cache, &k);
+    if (cbin != NULL) {
+      strb_appendb(bin, cbin);
+      return GA_NO_ERROR;
+    }
+  }
+
+  err = call_compiler(ctx, src, &ptx, log);
+  if (err != GA_NO_ERROR) return err;
+  err = make_bin(ctx, &ptx, bin, log);
+  if (err != GA_NO_ERROR) return err;
+  if (ctx->disk_cache) {
+    pk = memdup(&k, sizeof(k));
+    if (pk == NULL)
+      return GA_NO_ERROR;
+    cbin = strb_alloc(bin->l);
+    if (cbin == NULL) {
+      free(pk);
+      return GA_NO_ERROR;
```

Comment:
I would prefer a warning, but I think it isn't easy. So a memory error seem more appropriate as this is the only cause of this that I can think of.